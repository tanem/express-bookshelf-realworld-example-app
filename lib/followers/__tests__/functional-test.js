'use strict';

const chance = require('chance').Chance('follow');
const request = require('supertest');
const app = require('../../app');

const {followers, users} = app.locals.services;

describe('followers', () => {
  let user;
  let follower;

  beforeEach(async () => {
    user = await users.create({
      bio: chance.profession(),
      email: chance.email(),
      image: chance.avatar(),
      password: chance.word({length: 8}),
      username: chance.word(),
    });
    follower = await users.create({
      bio: chance.profession(),
      email: chance.email(),
      image: chance.avatar(),
      password: chance.word({length: 8}),
      username: chance.word(),
    });
    await followers.create({
      user: user.id,
      follower: follower.id,
    });
  });

  describe('POST /api/profiles/:username/follow', () => {
    describe('when the user is not authenticated', () => {
      test('should return an unauthorized error', async () => {
        const {body, status} = await request(app)
          .post(`/api/profiles/${follower.get('username')}/follow`)
          .set('Accept', 'application/json')
          .send();

        expect({body, status}).toMatchSnapshot();
      });
    });

    describe('when the user is authenticated', () => {
      describe('and the user to follow is not found', () => {
        test('should return a not found error', async () => {
          const {body, status} = await request(app)
            .post(`/api/profiles/invalid/follow`)
            .set('Accept', 'application/json')
            .set('Authorization', `Token ${users.generateJWT(user)}`)
            .send();

          expect({body, status}).toMatchSnapshot();
        });
      });

      describe('and the user to follow is found', () => {
        test('should follow the user to follow', async () => {
          const {body, status} = await request(app)
            .post(`/api/profiles/${follower.get('username')}/follow`)
            .set('Accept', 'application/json')
            .set('Authorization', `Token ${users.generateJWT(user)}`)
            .send();

          expect({body, status}).toMatchSnapshot();
        });
      });
    });
  });

  describe('DELETE /api/profiles/:username/follow', () => {
    describe('when the user is not authenticated', () => {
      test('should return an unauthorized error', async () => {
        const {body, status} = await request(app)
          .delete(`/api/profiles/${follower.get('username')}/follow`)
          .set('Accept', 'application/json')
          .send();

        expect({body, status}).toMatchSnapshot();
      });
    });

    describe('when the user is authenticated', () => {
      describe('and the user to unfollow is not found', () => {
        test('should return a not found error', async () => {
          const {body, status} = await request(app)
            .delete(`/api/profiles/invalid/follow`)
            .set('Accept', 'application/json')
            .set('Authorization', `Token ${users.generateJWT(user)}`)
            .send();

          expect({body, status}).toMatchSnapshot();
        });
      });

      describe('and the user to unfollow is found', () => {
        test('should unfollow the user to unfollow', async () => {
          const {body, status} = await request(app)
            .delete(`/api/profiles/${user.get('username')}/follow`)
            .set('Accept', 'application/json')
            .set('Authorization', `Token ${users.generateJWT(follower)}`)
            .send();

          expect({body, status}).toMatchSnapshot();
        });
      });
    });
  });
});
