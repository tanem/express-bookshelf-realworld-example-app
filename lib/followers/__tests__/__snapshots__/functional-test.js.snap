// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`followers DELETE /api/profiles/:username/follow when the user is authenticated and the user to unfollow is found should unfollow the user to unfollow 1`] = `
{
  "body": {
    "profile": {
      "bio": "Account Executive",
      "following": false,
      "image": "//www.gravatar.com/avatar/042d96a198ea28c8f54e055a3679e9cd",
      "username": "wut",
    },
  },
  "status": 200,
}
`;

exports[`followers DELETE /api/profiles/:username/follow when the user is authenticated and the user to unfollow is not found should return a not found error 1`] = `
{
  "body": {
    "errors": {
      "error": {
        "data": null,
        "isBoom": true,
        "isServer": false,
        "message": "Not Found",
        "output": {
          "headers": {},
          "payload": {
            "error": "Not Found",
            "message": "Not Found",
            "statusCode": 404,
          },
          "statusCode": 404,
        },
      },
      "message": "Not Found",
    },
  },
  "status": 404,
}
`;

exports[`followers DELETE /api/profiles/:username/follow when the user is not authenticated should return an unauthorized error 1`] = `
{
  "body": {
    "errors": {
      "error": {
        "data": null,
        "isBoom": true,
        "isServer": false,
        "message": "Unauthorized",
        "output": {
          "headers": {},
          "payload": {
            "error": "Unauthorized",
            "message": "Unauthorized",
            "statusCode": 401,
          },
          "statusCode": 401,
        },
      },
      "message": "Unauthorized",
    },
  },
  "status": 401,
}
`;

exports[`followers POST /api/profiles/:username/follow when the user is authenticated and the user to follow is found should follow the user to follow 1`] = `
{
  "body": {
    "profile": {
      "bio": "Chief Medical Officer",
      "following": true,
      "image": "//www.gravatar.com/avatar/4bdd6542a6dbc1d204b4cdab46221aed",
      "username": "soh",
    },
  },
  "status": 200,
}
`;

exports[`followers POST /api/profiles/:username/follow when the user is authenticated and the user to follow is not found should return a not found error 1`] = `
{
  "body": {
    "errors": {
      "error": {
        "data": null,
        "isBoom": true,
        "isServer": false,
        "message": "Not Found",
        "output": {
          "headers": {},
          "payload": {
            "error": "Not Found",
            "message": "Not Found",
            "statusCode": 404,
          },
          "statusCode": 404,
        },
      },
      "message": "Not Found",
    },
  },
  "status": 404,
}
`;

exports[`followers POST /api/profiles/:username/follow when the user is not authenticated should return an unauthorized error 1`] = `
{
  "body": {
    "errors": {
      "error": {
        "data": null,
        "isBoom": true,
        "isServer": false,
        "message": "Unauthorized",
        "output": {
          "headers": {},
          "payload": {
            "error": "Unauthorized",
            "message": "Unauthorized",
            "statusCode": 401,
          },
          "statusCode": 401,
        },
      },
      "message": "Unauthorized",
    },
  },
  "status": 401,
}
`;
